AWS EFS CSI driver (#785)
giantswarm/docs: src/content/guides/using-persistent-volumes-on-aws-with-efs-csi-driver/index.md

---
title: "Using persistent volumes with the EFS CSI driver on AWS"
description: "Tutorial on how to use persistent volumes with the EFS CSI driver on a cluster running on Amazon Web Services with EFS"
type: page
weight: 50
tags: ["tutorial"]
user_questions:
  - How do I install the EFS CSI driver?
  - How do I provision an EFS instance on AWS?
  - How do I test the EFS CSI storage class?
  - How do I use EFS CSI Volumes?
  - What auto-provisioner for AWS do you recommend? EFS? EBS?
owner:
  - https://github.com/orgs/giantswarm/teams/team-firecracker
---

> # Using Persistent Volumes on AWS with EFS

Reviewer: ```suggestion
# Using persistent volumes on AWS with EFS
```

Recently we decided to use [sentence style capitalization](https://docs.microsoft.com/de-de/style-guide/text-formatting/using-type/use-sentence-style-capitalization) in our docs, including headlines.

Also technical terms like e. g. "persistent volumes" and "availability zones" should be spelled in lower case. Exception: Code snippets/Literals where the exact case is important, e. g.

> ... the `PersistentVolumeClaim` resource ...

We haven't executed this decision in all out docs yet. However it would be nice if you could do it for this new article.

Author: This make sense 👍🏻

- All Kubernetes Persistent Volumes will be stored in the same EFS instance. You can deploy multiple provisioners per cluster, each having its own storage-class and EFS instance.
- [EFS throughput](https://docs.aws.amazon.com/efs/latest/ug/performance.html) need to be set up accordingly in order to not have performance issues. We only recommend Provisioned Throughput, and if you need high performance you will need EBS.
- EFS backups are done with [AWS Backup](https://aws.amazon.com/backup/) and it does not have the snapshot feature of EBS.
- You cannot limit the amount of data stored in an EFS volume. The requested value in Kubernetes is ignored.
- EFS mount targets are limited to 1 subnet per Availability Zone. Each NodePool will create a different subnet per AZ, plan accordingly.

## Provision an EFS instance on AWS

**Note:** Currently only static provisioning is supported. This means an AWS EFS file system needs to be created manually on AWS first. After that it can be mounted inside a container as a volume using the driver.

Before installing the provisioner in Kubernetes we will need to create the EFS instance in the same AWS account:

> 1. Open the [AWS management](https://aws.amazon.com/console/) console in the account your cluster is located.

Reviewer: ```suggestion
1. Open the [AWS management](https://aws.amazon.com/console/) console for the AWS account holding your cluster resources.
```

> 2. Select EFS from the services list.

Reviewer: ```suggestion
2. From the `Services` menu, select `EFS`.
```

Putting the terms in the order the user will look for them should make them easier to find. Also backticks make them stand out as literals.

> 3. Create a new file system and select the VPC where your cluster is located.

Reviewer: How do I identify the VPC? Can that be done by name? If not, please add  step beforehand, explaining how to find the VPC ID.

Author: Added info to pick up the cluster id for the right VPC.

> 4. Select the Availability Zone with the subnets your instances are located on and the security-groups of your node pools.

Reviewer: Is it obvious how to identify these AZs, subnets, instances and security groups? If not, plese explain this in a bit more detail.

Author: Added info to pick up the cluster id for the right subnets and security groups.

> 5. Choose the throughput and performance mode, no file system policy or access points are needed.

Reviewer: Attempt to clarify the second part:

```suggestion
5. Choose the throughput and performance mode. Setting a file system policy or access points is optional.
```

Is it obvious which choices are the right ones for throughput and performance mode? If not, please explain a bit more.

Author: I added links to for the options and recommended settings.

> 6. Create the instance and note the EFS instance id.

Reviewer: ```suggestion
6. Create the instance and note the EFS instance ID.
```

Reviewer: @njuettner Looks like this one got lost.

## Installing the EFS CSI driver

> To install the EFS CSI driver you will need to follow these steps:

Reviewer: Some clarification

```suggestion
To install the EFS CSI driver in the workload cluster, you will need to follow these steps:
```

1. Access the [web interface](/reference/web-interface/) and select the cluster on which you want to install the EFS CSI driver.
2. Open the Apps tab.
3. Click the Install App button
4. Select the Giant Swarm Playground catalog.

> 5. Select the App named aws-efs-csi-driver.

Reviewer: ```suggestion
5. Select the App named `aws-efs-csi-driver `.
```

Feel free to wrap menu and button names in backticks.

6. Click the Configure & Install button. Make sure that the correct cluster is selected.
7. Click the Install App button.

## Storage classes

> Once you installed the `aws-efs-csi-driver` from _Giant Swarm Playground_ catalog, Kubernetes cluster will have a storage class `efs-csi` deployed.

Reviewer: ```suggestion
Once you installed the `aws-efs-csi-driver` from _Giant Swarm Playground_ catalog, you workload cluster will have a storage class `efs-csi` deployed.
```

## Deploy a sample application

> In order to see if all works out we need to create some manifests:

Reviewer: ```suggestion
In order to verify that the EFS CSI driver works as expected, we suggest to deploy a test workload based on the following manifests.
```

> First we apply a PersistentVolume with a capacity of 5 GB storage, here you need to provide EFS instance id:

Reviewer: ```suggestion
First we apply a persistent volume with a capacity of 5 GB storage. Make sure to apply your EFS instance ID (noted before) as `.spec.csi. volumeHandle` value!
```

```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  name: efs-pv
spec:
  capacity:
    storage: 5Gi
  volumeMode: Filesystem
  accessModes:
    - ReadWriteMany
  persistentVolumeReclaimPolicy: Retain
  storageClassName: efs-csi
  csi:
    driver: efs.csi.aws.com
    volumeHandle: fs-xxxxxxxx
```

> **Note:** Storage capacity is a required field, you need to specify a valid value but is not used when creating the file system. It's important to note EFS is an elastic file system, it does not enforce any file system capacity limits.

Reviewer: ```suggestion
**Note:** `.spec.storage.capacity` is a required field. You must specify a valid value, however that value is not used when creating the file system. It's important to note that EFS is an elastic file system, which does not enforce any file system capacity limits.
```

> Additionally we need a PerstentVolumeClaim with a request of 5 GB storage:

Reviewer: ```suggestion
Additionally we need a persistent volume claim with a storage request matching the size of the persistent volume defined above:
```

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: efs-claim
spec:
  accessModes:
    - ReadWriteMany
  storageClassName: efs-csi
  resources:
    requests:
      storage: 5Gi
```

> And finally, we apply a Pod which writes every 5 seconds the current date in UTC time in a file located in our EFS storage mount:

Reviewer: ```suggestion
And finally, we create a pod which repeatedly writes the current date into a file located in our EFS storage mount:
```

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: efs-app
spec:
  containers:
  - name: linux
    image: amazonlinux:2
    command: ["/bin/sh"]
    args: ["-c", "while true; do echo $(date -u) >> /efs-data/out.txt; sleep 5; done"]
    volumeMounts:
    - name: efs-storage
      mountPath: /efs-data
  volumes:
  - name: efs-storage
    persistentVolumeClaim:
      claimName: efs-claim
```

> **Warning:**
> By default, new EFS file systems are owned by root:root. You might need to change file system permissions if your container is not running as root.

Reviewer: ```suggestion
**Warning:** By default, new EFS file systems are owned by root:root. You might need to change file system permissions if your container is not running as root.
```

Will readers know how to do that, or can we make it easier for them?

Author: I added more context how you can change permissions by using EFS access points.

- [EFS throughput](https://docs.aws.amazon.com/efs/latest/ug/performance.html) need to be set up accordingly in order to not have performance issues. We only recommend Provisioned Throughput, and if you need high performance you will need EBS.
- EFS backups are done with [AWS Backup](https://aws.amazon.com/backup/) and it does not have the snapshot feature of EBS.
- You cannot limit the amount of data stored in an EFS volume. The requested value in Kubernetes is ignored.
- EFS mount targets are limited to 1 subnet per availability zone. Each NodePool will create a different subnet per AZ, plan accordingly.

## Provision an EFS instance on AWS

**Note:** Currently only static provisioning is supported. This means an AWS EFS file system needs to be created manually on AWS first. After that it can be mounted inside a container as a volume using the driver.

Before installing the provisioner in Kubernetes we will need to create the EFS instance in the same AWS account:

1. Open the [AWS management](https://aws.amazon.com/console/) console for the AWS account holding your cluster resources.
2. From the `Services` menu, select `EFS`.

> 3. Create a new file system and select the VPC where your cluster is located. VPC can be identified by your cluster ID.

Reviewer: ```suggestion
3. Create a new file system and select the VPC where your cluster is located. The VPC can be identified by your cluster ID.
```

  csi:
    driver: efs.csi.aws.com
    volumeHandle: fs-xxxxxxxx
```

**Note:** `.spec.storage.capacity` is a required field. You must specify a valid value, however that value is not used when creating the file system. It's important to note that EFS is an elastic file system, which does not enforce any file system capacity limits.

Additionally we need a persistent volume claim with a storage request matching the size of the persistent volume defined above:

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: efs-claim
spec:
  accessModes:
    - ReadWriteMany
  storageClassName: efs-csi
  resources:
    requests:
      storage: 5Gi
```

And finally, we create a pod which repeatedly writes the current date into a file located in our EFS storage mount:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: efs-app
spec:
  containers:
  - name: linux
    image: amazonlinux:2
    command: ["/bin/sh"]
    args: ["-c", "while true; do echo $(date -u) >> /efs-data/out.txt; sleep 5; done"]
    volumeMounts:
    - name: efs-storage
      mountPath: /efs-data
  volumes:
  - name: efs-storage
    persistentVolumeClaim:
      claimName: efs-claim
```

**Warning:**

> By default, new EFS file systems are owned by root:root. You might need to change file system permissions if your container is not running as root. Exposing separate data stores with independent ownership and permissions you can check [EFS access points](https://docs.aws.amazon.com/efs/latest/ug/efs-access-points.html) to get familiar.

Reviewer: ```suggestion
By default, new EFS file systems are owned by root:root. You might need to change file system permissions if your container is not running as root. To learn about exposing separate data stores with independent ownership and permissions, check the AWS guide on [working with Amazon EFS Access Points](https://docs.aws.amazon.com/efs/latest/ug/efs-access-points.html).
```
